draft/ZEP0010.md

```diff
@@ -0,0 +1,440 @@
+
+Metadata is introduced in the ``example.multiscale-arrays``
+extension which allows encoding a relationship between multiple arrays at the
+group level. This defines a "multiscale pyramid" of arrays which is
+a common idiom in both the geospatial and bioimaging uses of Zarr.
+Implementations may choose to return a different subclass or backend when
+detecting such metadata. In this case, a "datatree" which allows similar
+operations on all levels of the pyramid might be preferred.
+
+#### Tiered storage (group)
+
+```javascript
+{
+    "zarr_format": 3,
+    "node_type": "group",
+    ...,
+    "extensions": [
+        {
+            "name": "example.tiered-storage",
+            "must_understand": false,
+            "configuration": {
+                "slow-arrays": [
+                    "path/to/array/1"
+                ]
+            }
+        }
+    ],
+}
+```
+
+Related to the multiscales example above, an ``example.tiered-storage``
+extension could identify arrays within a group which have been put on
+slower or even archived filesystems which will encourage more overhead
+and potentially costs if they are accessed. An implementation might
+warn users before opening the array.
+
+### Application to sub-nodes
+
+This ZEP does not try to define the behavior for application to sub-nodes
+itself, but defers this to actual extensions.
+
+Conceptually, we propose that extensions defined on groups may be valid for
+their child nodes. However, the details of how an implementation should
+identify which extensions are active within an hierarchy are unclear. Relying
+on traversing the hierarchy towards the root node is undesirable from a
+performance point of view.
+
+As a workaround, extension authors can choose to write *some* metadata within
+the contained subgroups and arrays to make this easier. Options for what
+this metadata could be include:
+
+1. A copy of the metadata
+
+```javascript
+{
+  "extensions":  [
+    {
+      "name": "example.my-extension",
+      "configuration": { ... full copy of the metadata ...}
+    }
+  ]
+
+}
+```
+
+2. A reference to the metadata as part of the extension itself
+
+```javascript
+{
+  "extensions":  [
+    {
+      "name": "example.my-extension",
+      "configuration": {
+        "reference": "../.."
+      }
+    }
+  ]
+
+}
+```
+
+3. A complimentary reference extension
+
+```javascript
+{
+  "extensions":  [
+    {
+      "name": "example.my-extension-ref",
+      "configuration": {
+        "reference": "../.."
+      }
+    }
+  ]
+
+}
+```
+
+4. A shared or even core reference extension
+
+```javascript
+{
+  "extensions":  [
+    {
+      "name": "example.parent-ref",
+      "configuration": {
+        "reference": "../.."
+      }
+    }
+  ]
+}
+```
+
+As further experience is gained by the community of extension authors,
+one or more of these methods may be adopted into the core spec.
+
+### Alternatives for the `extensions` extension point
+
+The current design allows having the same
```

Comment:
I don't understand "This MAY happen as
part of the core spec adopting functionality of an extension.". Can you provide more details about the motivation for this warning?

Reply:
Let's use the "offset" example. Currently, someone could grab the zarr-extensions name and start building that extension. In the future, another ZEP could codify the name "offset" as top-level metadata, independently of the fact that it would then already be an extension.